### PR TITLE
Add unit tests for `simulate_tree_mcm`

### PR DESCRIPTION
**Pull request by obliviateandsurrender**
**Context:** Adds unit tests for the functionality to simulate mid-circuit measurements natively.

**Description of the Change:** 

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:** [sc-71875]

### CODE REVIEW

File: pennylane/devices/qubit/simulate.py [hunk 1]
```diff
@@ -922,7 +922,7 @@ def _(original_measurement: ExpectationMP, measures):  # pylint: disable=unused-
     for v in measures.values():
         if not v[0] or v[1] is tuple():
             continue
-        cum_value += v[0] * v[1]
+        cum_value += v[0] * qml.math.squeeze(v[1])
```
Comment (astralcai) [inline pennylane/devices/qubit/simulate.py:925]:
```suggestion
        cum_value += qml.math.multiply(v[0], v[1])
```

File: pennylane/devices/qubit/simulate.py [hunk 2]
```diff
@@ -935,14 +935,14 @@ def _(original_measurement: ProbabilityMP, measures):  # pylint: disable=unused-
...
```
Comment (astralcai) [inline pennylane/devices/qubit/simulate.py:938]:
```suggestion
        cum_value += qml.math.multiply(v[0], v[1])
```

File: tests/devices/qubit/test_simulate.py [hunk 3]
```diff
@@ -1178,3 +1184,228 @@ def test_qinfo_tf(self):
...
+    # pylint: disable=too-many-arguments
+    @pytest.mark.parametrize("shots", [None, 5500])
+    @pytest.mark.parametrize("postselect", [None, 0])
+    @pytest.mark.parametrize("reset", [False, True])
+    @pytest.mark.parametrize("measure_f", [qml.counts, qml.expval, qml.probs, qml.sample, qml.var])
+    @pytest.mark.parametrize(
+        "meas_obj",
+        [qml.Y(0), [1], [1, 0], "mcm", "composite_mcm", "mcm_list"],
+    )
+    def test_simple_dynamic_circuit(self, shots, measure_f, postselect, reset, meas_obj):
```
Comment (astralcai) [inline tests/devices/qubit/test_simulate.py:1227]: Why is this included in this PR?

File: tests/devices/qubit/test_simulate.py [hunk 4]
```diff
@@ -1178,3 +1184,228 @@ def test_qinfo_tf(self):
 
         grad5 = grad_tape.jacobian(results[5], phi)
         assert qml.math.allclose(grad5, expected_grads[5])
+
+
+class TestMidMeasurements:
+    """Tests for simulating scripts with mid-circuit measurements using the ``simulate_tree_mcm``."""
+
+    @pytest.mark.parametrize("val", [0, 1])
+    def test_basic_mid_meas_circuit(self, val):
```
Comment (astralcai) [inline tests/devices/qubit/test_simulate.py:1193]:
```suggestion
    @pytest.mark.unit
    @pytest.mark.parametrize("val", [0, 1])
    def test_basic_mid_meas_circuit(self, val):
```
Can you add the test level markers to the rest of the tests as well?

File: tests/devices/qubit/test_simulate.py (hunk 3 above)
Comment (obliviateandsurrender) [reply tests/devices/qubit/test_simulate.py:1227]: I added it here because I am removing a similar one from `default_qubit_native_mcm` in this [PR](https://github.com/PennyLaneAI/pennylane/pull/6246)

File: tests/devices/qubit/test_simulate.py [hunk 5]
```diff
@@ -1178,3 +1184,232 @@ def test_qinfo_tf(self):
...
+        qs_with_mid_meas = qml.tape.QuantumScript(
+            [
+                qml.Hadamard(0),
+                qml.Hadamard(1),
+                qml.CZ([0, 1]),
+                qml.CNOT([0, 2]),
+                qml.CNOT([2, 3]),
+                qml.CZ([1, 3]),
+                qml.Toffoli([3, 2, 0]),
+                (m0 := qml.measure(0)).measurements[0],
```
Comment (mudit2812) [inline tests/devices/qubit/test_simulate.py:1322]: This is so much nicer than what I used in tests that I wrote 😍

**Review by mudit2812 — CHANGES_REQUESTED** [review]
Looks good. Many of the tests don't verify the numerical correctness of the tests. For that, we would probably need a strategy similar to #6124, which is blocked by discussions.

**Comment by astralcai** [comment]
> Looks good. Many of the tests don't verify the numerical correctness of the tests. For that, we would probably need a strategy similar to https://github.com/PennyLaneAI/pennylane/pull/6124, which is blocked by discussions.

Seconding Mudit here, can we verify the numerical correctness where it makes sense?

**Comment by obliviateandsurrender** [comment]
@mudit2812 and @astralcai I have added numerical assertions to `test_tree_traversal_interface_mcm` and `test_approx_dynamic_mid_meas_circuit`.

File: tests/devices/qubit/test_simulate.py [hunk 6]
```diff
@@ -1178,3 +1198,371 @@ def test_qinfo_tf(self):
...
+        tree_stack = TreeTraversalStack(max_depth)
+
+        assert tree_stack.counts.count(None) == max_depth
+        assert tree_stack.probs.count(None) == max_depth
+        assert tree_stack.results_0.count(None) == max_depth
+        assert tree_stack.results_1.count(None) == max_depth
+        assert tree_stack.states.count(None) == max_depth
+
+    @pytest.mark.unit
+    def test_full_prune_empty_methods(self):
```
Comment (mudit2812) [inline tests/devices/qubit/test_simulate.py:1222]:
```suggestion

@pytest.mark.unit
class TestTreeTraversalStack:
    """Unit tests for TreeTraversalStack"""

    @pytest.mark.parametrize(
        "max_depth",
        [0, 1, 10, 100],
    )
    def test_init_with_depth(self, max_depth):
        """Test that TreeTraversalStack is initialized correctly with given ``max_depth``"""
        tree_stack = TreeTraversalStack(max_depth)

        assert tree_stack.counts.count(None) == max_depth
        assert tree_stack.probs.count(None) == max_depth
        assert tree_stack.results_0.count(None) == max_depth
        assert tree_stack.results_1.count(None) == max_depth
        assert tree_stack.states.count(None) == max_depth

    def test_full_prune_empty_methods(self):
```

File: tests/devices/qubit/test_simulate.py [hunk 7]
```diff
@@ -1178,3 +1198,371 @@ def test_qinfo_tf(self):
...
+    # pylint: disable=too-many-arguments
+    @pytest.mark.unit
+    @pytest.mark.parametrize("shots", [None, 5500])
+    @pytest.mark.parametrize("postselect", [None, 0])
+    @pytest.mark.parametrize("reset", [False, True])
+    @pytest.mark.parametrize("measure_f", [qml.counts, qml.expval, qml.probs, qml.sample, qml.var])
+    @pytest.mark.parametrize(
+        "meas_obj", [qml.Y(0), [1], [1, 0], "mcm", "composite_mcm", "mcm_list"]
+    )
+    def test_simple_dynamic_circuit(self, shots, measure_f, postselect, reset, meas_obj):
```
Comment (mudit2812) [inline tests/devices/qubit/test_simulate.py:1298]: Instead of marking the individual tests as `unit`, you can mark the whole test class.

**Review by astralcai — APPROVED** [review]
LGTM

File: tests/devices/qubit/test_simulate.py [hunk 8]
```diff
@@ -1178,3 +1198,371 @@ def test_qinfo_tf(self):
...
+        assert len(mcms) == n_circs
+
+        split_circs = split_circuit_at_mcms(qscript)
+        assert len(split_circs) == n_circs + 1
+        for circ in split_circs:
+            assert not [o for o in circ.operations if isinstance(o, qml.measurements.MidMeasureMP)]
+
+        res = simulate_tree_mcm(qscript, postselect_mode="fill-shots")
+        assert len(res[0]) == 20
+        assert isinstance(res[1], dict) and sum(list(res[1].values())) == 20
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply tests/devices/qubit/test_simulate.py]: This was meant to test that `simulate_tree_mcm` can handle sufficiently large numbers of `mcms`. I brought it here from `default_qubit_native_mcm` and extended it to test the auxiliary functions the tree-traversal algorithm uses for finding `mcms` that require post-processing and for splitting the tape on those particular `mcms`.

File: tests/devices/qubit/test_simulate.py [hunk 9]
```diff
@@ -1178,3 +1198,371 @@ def test_qinfo_tf(self):
...
+        "postselect_mode", [None, "hw-like", "pad-invalid-samples", "fill-shots"]
+    )
+    def test_tree_traversal_interface_mcm(self, ml_framework, postselect_mode):
+        """Test that tree traversal works numerically with different interfaces"""
+        # pylint:disable = singleton-comparison, import-outside-toplevel
+        if ml_framework == "tensorflow":
+            import tensorflow as tf
+
+            tf.experimental.numpy.experimental_enable_numpy_behavior()
+
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply tests/devices/qubit/test_simulate.py:1451]: Otherwise, the test for `tensorflow` interface fails because `ravel` can only be done when I enable `numpy` behavior. I didn't dig deep into what part of workflow was causing, but I suspect it using turning the `reset=True` for `m0`. I'll add a line for disabling at the end.

[the remark this answers is not in the document]
Comment (mudit2812) [reply tests/devices/qubit/test_simulate.py:1451]: We should probably not use `array.ravel` in that case. We can just use `qml.math.reshape` or `qml.math.ravel` wherever the error is coming from. I see this is a bug in the code rather than a bug in the test.

File: tests/devices/qubit/test_simulate.py [hunk 10]
```diff
@@ -1178,3 +1198,371 @@ def test_qinfo_tf(self):
...
+                prob_dist1, prob_dist2 = np.zeros(2**n_wires), np.zeros(2**n_wires)
+                for prob, rs in zip([prob_dist1, prob_dist2], [rs1, rs2]):
+                    index, count = np.unique(
+                        np.packbits(rs, axis=1, bitorder="little").squeeze(), return_counts=True
+                    )
+                    prob[index] = count
+
+            assert qml.math.allclose(
+                sp.stats.entropy(prob_dist1 + 1e-12, prob_dist2 + 1e-12), 0.0, atol=5e-2
+            )
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply tests/devices/qubit/test_simulate.py:1438]: The motivation here is to calculate the KL divergence (KLD) between the two probability distributions computed from the samples obtained from the two tapes. It gives a measure that shows how different two distributions.

File: tests/devices/qubit/test_simulate.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply tests/devices/qubit/test_simulate.py]: Fixed it.

**Review by mudit2812 — APPROVED** [review]
